src/storage.py

```diff
@@ -0,0 +1,283 @@
+#!/usr/bin/env python3
+
+# Copyright 2024 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Handle Juju Storage Events."""
+
+import json
+import logging
+from subprocess import CalledProcessError, run
+
+import ops_sunbeam.guard as sunbeam_guard
+from ops.charm import CharmBase, StorageAttachedEvent, StorageDetachingEvent
+from ops.framework import Object, StoredState
+from ops.model import ActiveStatus, MaintenanceStatus
+from tenacity import retry, stop_after_attempt, wait_fixed
+
+import microceph
+
+logger = logging.getLogger(__name__)
+
+
+class StorageHandler(Object):
+    """The Storage class manages the Juju storage events.
+
+    Observes the following events:
+    1) *_storage_attached
+    2) *_storage_detaching
+    """
+
+    name = "storage"
+
+    # storage directive names
+    standalone = "osd-standalone"
+    dedicated_db = "osd-dedicated-db"
+
+    charm = None
+    """
+..        osd_data: dict of dicts with int (osd num) key
+            disk_by_id: OSD disk by id
+            disk: OSD disk path
+            db: db disk path
+    """
+    _stored = StoredState()
+
+    def __init__(self, charm: CharmBase, name="storage"):
+        super().__init__(charm, name)
+        self._stored.set_default(osd_data={})
+        self.charm = charm
+        self.name = name
+
+        # Attach handlers
+        self.framework.observe(
+            charm.on[self.standalone.replace("-", "_")].storage_attached,
+            self._on_osd_standalone_attached,
+        )
+        for directive in [self.dedicated_db.replace("-", "_"), "db"]:
+            self.framework.observe(charm.on[directive].storage_attached, self._on_attached)
+
+        # OSD Detaching handlers.
+        for directive in [
+            self.standalone.replace("-", "_"),
+            self.dedicated_db.replace("-", "_"),
+            "db",
+        ]:
+            self.framework.observe(
+                charm.on[directive].storage_detaching, self._on_storage_detaching
+            )
+
+    # storage event handlers
+
+    def _on_attached(self, event: StorageAttachedEvent):
```

Comment:
How are the device paths presented to the charm? If not already referring to a stable path I'd suggest to normalize the device paths for this function similar to what microceph does in setStablePath https://github.com/canonical/microceph/blob/b95919779485dd6adc2dd0761730a273f44e9a77/microceph/ceph/osd.go#L248

Reply:
Ack, have modified the logic to always refer using storage name (directive/index) which is unique for each storage device. Paths are fetched only when required, this ensures stored state does not become inconsistent with reality if disknames are changed during reboot.